File: src/execution-policy.ts
import { formatSize } from './utils.js';

export interface AgentCLIExecutionPolicy {
  maxMaterializedBytes?: number;
}

export type ResolvedAgentCLIExecutionPolicy = Readonly<AgentCLIExecutionPolicy>;

export interface MaterializedByteLimit {
  limit: number;
  actual: number;
}

const EMPTY_EXECUTION_POLICY: ResolvedAgentCLIExecutionPolicy = Object.freeze({});

export function resolveExecutionPolicy(policy?: AgentCLIExecutionPolicy): ResolvedAgentCLIExecutionPolicy {
  if (!policy) {
    return EMPTY_EXECUTION_POLICY;
  }

  const resolved: AgentCLIExecutionPolicy = {};
  validateLimit('executionPolicy.maxMaterializedBytes', policy.maxMaterializedBytes);

  if (policy.maxMaterializedBytes !== undefined) {
    resolved.maxMaterializedBytes = policy.maxMaterializedBytes;
  }

  return Object.freeze(resolved);
}

export function checkMaterializedByteLimit(
  policy: ResolvedAgentCLIExecutionPolicy,
  actual: number,
): MaterializedByteLimit | null {
  const limit = policy.maxMaterializedBytes;
  if (limit === undefined || actual <= limit) {
    return null;
  }

  return { limit, actual };
}

export function formatMaterializedLimitMessage(
  commandName: string,
  subject: string,
  actual: number,
  limit: number,
): string {
  return `${commandName}: input exceeds max materialized size (${formatSize(actual)} > ${formatSize(limit)}): ${subject}`;
}

function validateLimit(label: string, value: number | undefined): void {
  if (value === undefined) {
    return;
  }
  if (!Number.isInteger(value) || value < 0) {
    throw new Error(`${label} must be a non-negative integer`);
  }
}

File: python/src/onetool/execution_policy.py
from __future__ import annotations

from dataclasses import dataclass

from .utils import format_size


@dataclass(frozen=True, slots=True)
class AgentCLIExecutionPolicy:
    max_materialized_bytes: int | None = None


ResolvedAgentCLIExecutionPolicy = AgentCLIExecutionPolicy


@dataclass(frozen=True, slots=True)
class MaterializedByteLimit:
    limit: int
    actual: int


EMPTY_EXECUTION_POLICY = AgentCLIExecutionPolicy()


def resolve_execution_policy(
    policy: AgentCLIExecutionPolicy | None = None,
) -> ResolvedAgentCLIExecutionPolicy:
    if policy is None:
        return EMPTY_EXECUTION_POLICY

    _validate_limit("executionPolicy.maxMaterializedBytes", policy.max_materialized_bytes)
    return AgentCLIExecutionPolicy(max_materialized_bytes=policy.max_materialized_bytes)


def check_materialized_byte_limit(
    policy: ResolvedAgentCLIExecutionPolicy,
    actual: int,
) -> MaterializedByteLimit | None:
    limit = policy.max_materialized_bytes
    if limit is None or actual <= limit:
        return None
    return MaterializedByteLimit(limit=limit, actual=actual)


def format_materialized_limit_message(
    command_name: str,
    subject: str,
    actual: int,
    limit: int,
) -> str:
    return (
        f"{command_name}: input exceeds max materialized size "
        f"({format_size(actual)} > {format_size(limit)}): {subject}"
    )


def _validate_limit(label: str, value: int | None) -> None:
    if value is None:
        return
    if not isinstance(value, int) or value < 0:
        raise ValueError(f"{label} must be a non-negative integer")
